### PR TITLE
 Persistent volumes for pods — ZK data migration 

### PR DESCRIPTION
**Pull request by meln1k**
Summary: added a migration to move residency from the task level to the instance level

JIRA issues: MARATHON_EE-1785

Results of the manual test:
`[zk: localhost:2181(CONNECTED) 1] get /marathon/state/instance/8/postgres.marathon-3b4a55ee-f547-11e7-9abc-22ca52e973a9`
```json
{
   "instanceId":{
      "idString":"postgres.marathon-3b4a55ee-f547-11e7-9abc-22ca52e973a9"
   },
   "agentInfo":{
      "host":"ubuntu",
      "agentId":"197fdeb5-94a8-44ac-90d6-f44de4a6fc6f-S0",
      "attributes":[

      ]
   },
   "tasksMap":{
      "postgres.3b4a55ee-f547-11e7-9abc-22ca52e973a9.304":{
         "taskId":"postgres.3b4a55ee-f547-11e7-9abc-22ca52e973a9.304",
         "reservation":{
            "volumeIds":[
               {
                  "runSpecId":"/postgres",
                  "containerPath":"pgdata",
                  "uuid":"3b4a55ed-f547-11e7-9abc-22ca52e973a9",
                  "persistenceId":"postgres#pgdata#3b4a55ed-f547-11e7-9abc-22ca52e973a9"
               }
            ],
            "state":{
               "name":"suspended",
               "timeout":null
            }
         },
         "status":{
            "stagedAt":"2018-01-09T15:06:35.523Z",
            "mesosStatus":"CjMKMXBvc3RncmVzLjNiNGE1NWVlLWY1NDctMTFlNy05YWJjLTIyY2E1MmU5NzNhOS4zMDQQAyIwQWJub3JtYWwgZXhlY3V0b3IgdGVybWluYXRpb246IHVua25vd24gY29udGFpbmVyKikKJzE5N2ZkZWI1LTk0YTgtNDRhYy05MGQ2LWY0NGRlNGE2ZmM2Zi1TMDFpOOSeNpXWQTozCjFwb3N0Z3Jlcy4zYjRhNTVlZS1mNTQ3LTExZTctOWFiYy0yMmNhNTJlOTczYTkuMzA0SAFQAVoQLb5WpFmSRXKnKFJhfJP/QA==",
            "condition":"Reserved",
            "networkInfo":{
               "hostName":"ubuntu",
               "hostPorts":[
                  31497
               ],
               "ipAddresses":[

               ]
            }
         },
         "runSpecVersion":"2018-01-09T14:13:12.774Z"
      }
   },
   "runSpecVersion":"2018-01-09T14:13:12.774Z",
   "state":{
      "condition":"Reserved",
      "since":"2018-01-09T15:06:35.574Z"
   },
   "unreachableStrategy":"disabled"
}
```
After the migration:
```json
{
   "instanceId":{
      "idString":"postgres.marathon-3b4a55ee-f547-11e7-9abc-22ca52e973a9"
   },
   "agentInfo":{
      "host":"ubuntu",
      "agentId":"197fdeb5-94a8-44ac-90d6-f44de4a6fc6f-S0",
      "attributes":[

      ]
   },
   "tasksMap":{
      "postgres.3b4a55ee-f547-11e7-9abc-22ca52e973a9.307":{
         "taskId":"postgres.3b4a55ee-f547-11e7-9abc-22ca52e973a9.307",
         "runSpecVersion":"2018-01-09T14:13:12.774Z",
         "status":{
            "stagedAt":"2018-01-10T10:42:46.340Z",
            "mesosStatus":"CjMKMXBvc3RncmVzLjNiNGE1NWVlLWY1NDctMTFlNy05YWJjLTIyY2E1MmU5NzNhOS4zMDcQAyIwQWJub3JtYWwgZXhlY3V0b3IgdGVybWluYXRpb246IHVua25vd24gY29udGFpbmVyKikKJzE5N2ZkZWI1LTk0YTgtNDRhYy05MGQ2LWY0NGRlNGE2ZmM2Zi1TMDEqjaCJe5XWQTozCjFwb3N0Z3Jlcy4zYjRhNTVlZS1mNTQ3LTExZTctOWFiYy0yMmNhNTJlOTczYTkuMzA3SAFQAVoQJtOvS7iOS7OLG6/I15XrPg==",
            "condition":"Reserved",
            "networkInfo":{
               "hostName":"ubuntu",
               "hostPorts":[
                  31497
               ],
               "ipAddresses":[

               ]
            }
         }
      }
   },
   "runSpecVersion":"2018-01-09T14:13:12.774Z",
   "state":{
      "condition":"Reserved",
      "since":"2018-01-10T10:42:46.545Z"
   },
   "unreachableStrategy":"disabled",
   "reservation":{
      "volumeIds":[
         {
            "runSpecId":"/postgres",
            "containerPath":"pgdata",
            "uuid":"3b4a55ed-f547-11e7-9abc-22ca52e973a9",
            "persistenceId":"postgres#pgdata#3b4a55ed-f547-11e7-9abc-22ca52e973a9"
         }
      ],
      "state":{
         "name":"suspended",
         "timeout":null
      }
   }
}
```

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+          case (reservation, instance) =>
+            val updatedInstance = instance.copy(reservation = Some(reservation))
+            instanceRepository.store(updatedInstance)
+        }
+        .runWith(Sink.ignore)
+    } getOrElse {
+      Future.successful(Done)
+    }
+  }
+}
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: nit: could you please add a new line here?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+@SuppressWarnings(Array("ClassNames"))
+class MigrationTo160(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer) extends StrictLogging {
+
+  def migrate(): Future[Done] = {
+    MigrationTo160.migrateReservations(instanceRepository, persistenceStore)(Environment(sys.env), ctx, mat)
+  }
+}
+
+object MigrationTo160 extends StrictLogging {
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit env: Environment, ctx: ExecutionContext, mat: Materializer): Future[Done] = {
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Could you please add a comment here giving an overview what this function does?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.MigrationCancelledException
+import mesosphere.marathon.core.async.ExecutionContexts
+import mesosphere.marathon.core.instance.{ Instance, Reservation }
+import mesosphere.marathon.core.instance.Instance.Id
+import mesosphere.marathon.core.storage.store.impl.cache.{ LazyCachingPersistenceStore, LazyVersionCachingPersistentStore, LoadTimeCachingPersistenceStore }
+import mesosphere.marathon.core.storage.store.{ IdResolver, PersistenceStore }
+import mesosphere.marathon.core.storage.store.impl.zk.{ ZkId, ZkPersistenceStore, ZkSerialized }
+import mesosphere.marathon.storage.migration.MigrationTo146.Environment
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: There is no need in it here, I believe.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    MigrationTo160.migrateReservations(instanceRepository, persistenceStore)(Environment(sys.env), ctx, mat)
+  }
+}
+
+object MigrationTo160 extends StrictLogging {
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit env: Environment, ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    val maybeStore: Option[ZkPersistenceStore] = {
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Could you please elaborate here in a comment why there is a need for a custom store.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    maybeStore.map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get(instanceId)
+        }
+        .mapConcat {
+          case Some(jsValue) =>
+            val instance = jsValue.as[Instance]
+            val maybeReservationJson = (jsValue \ "tasksMap" \\ "reservation").headOption
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: I think, it is worth mentioning that prior to 1.6.0, only PVs are supported only with apps, therefore `reservation` can only appear in app instances, and in such of apps, `taskMap` has only one KV pair.

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+import mesosphere.AkkaUnitTest
+import mesosphere.marathon.core.condition.Condition
+import mesosphere.marathon.core.instance.Instance.Id
+import mesosphere.marathon.core.instance.{ Instance, Reservation, TestInstanceBuilder }
+import mesosphere.marathon.core.storage.store.{ IdResolver, PersistenceStore }
+import mesosphere.marathon.core.storage.store.impl.zk.{ ZkId, ZkPersistenceStore, ZkSerialized }
+import mesosphere.marathon.core.task.Task
+import mesosphere.marathon.core.task.state.NetworkInfo
+import mesosphere.marathon.state._
+import mesosphere.marathon.storage.migration.MigrationTo146.Environment
```
Comment (ichernetsky) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala]: Maybe I am missing something, but why is it needed here?

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+import play.api.libs.json.{ JsValue, Json }
+import play.api.libs.json._
+
+import scala.concurrent.duration._
+import scala.concurrent.{ ExecutionContextExecutor, Future }
+
+class MigrationTo160Test extends AkkaUnitTest with GroupCreation with StrictLogging {
+
+  "Migration to 1.6.0" should {
+    "do nothing if env var is not configured" in new Fixture {
```
Comment (ichernetsky) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala]: The comment is misleading. I assume the idea is that nothing should be done, if there are no apps with PVs. This test covers the case when there is nothing stored ZK. I think it would be nice to cover the case in which nothing should be done, if there are apps without PVs and pods stored in ZK. What do you think?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 5 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Yes, but this code is also handling the case where we don't have any reservation at all, so I'd keep it as is.

Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Yes, I agree. I was not suggestion any change to the code. Instead, I was thinking that it would be a good idea to clarify why it happens the way it happens here in a comment. What do you think?

Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Sounds good!

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    * @param ctx
+    * @param mat
+    * @return
+    */
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    /**
+      * We're trying to find if we have a zookeeper store to make sure that migration will run in production environment
```
[the remark this answers is not in the document]
Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Also, is there a specific reason why `ZkPersistenceStore` is used, and not, say, just `PersistenceStore`? Why is it better to avoid running migrations in unit-tests?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+    maybeStore.map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get(instanceId)
+        }
+        .mapConcat {
+          case Some(jsValue) =>
+            val instance = jsValue.as[Instance]
+            //prior to 1.6.0, only PVs are supported only with apps,
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Nit:
```
            // Prior to Marathon 1.6.0, persistent volumes are supported only with apps,
            // therefore reservation objects can only appear in app instances, and since
            // an app has only one task by definition, there is only one KV pair in a taskMap
            // object.
            //
            // We use .headOption here to handle the case of apps with no persistent volumes.
```

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    /**
+      * We're trying to find if we have a ZooKeeper store because it provides objects as byte arrays and this
+      * makes serialization into json easier.
+      */
+    val maybeStore: Option[ZkPersistenceStore] = {
+
+      def findZkStore(ps: PersistenceStore[_, _, _]): Option[ZkPersistenceStore] = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: This is `@tailrec` isn't it?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    */
+  def migrateReservations(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit ctx: ExecutionContext, mat: Materializer): Future[Done] = {
+
+    logger.info("Starting reservations migration to 1.6.0")
+
+    /**
+      * We're trying to find if we have a ZooKeeper store because it provides objects as byte arrays and this
+      * makes serialization into json easier.
+      */
+    val maybeStore: Option[ZkPersistenceStore] = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: What happens if this is `None`?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+            findZkStore(ltcps.store)
+          case other =>
+            None
+        }
+      }
+
+      findZkStore(persistenceStore)
+    }
+
+    implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Why is this required?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            // We use .headOption here to handle the case of apps with no persistent volumes.
+            val maybeReservationJson = (jsValue \ "tasksMap" \\ "reservation").headOption
+
+            maybeReservationJson.map { reservationJson =>
+              reservationJson.as[Reservation] -> instance :: Nil
+            } getOrElse {
+              Nil
+            }
+
+          case _ => Nil
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Is the return type a list?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        }
+        .mapConcat { case (reservation, instance) =>
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: I thought we get a list from the `mapConcat` before. Why do we have a reservation and an instance here?

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+class MigrationTo160Test extends AkkaUnitTest with GroupCreation with StrictLogging {
+
+  "Migration to 1.6.0" should {
+    "do migration for instances with tasks with reservations" in new Fixture {
+      initMocks()
+      MigrationTo160.migrateReservations(instanceRepository, persistenceStore)(ctx, mat).futureValue
+      val targetInstance = instance.copy(reservation = Some(Reservation(Nil, Reservation.State.Launched)))
+      val targetInstance2 = instance2.copy(reservation = Some(Reservation(Nil, Reservation.State.Launched)))
+      val targetInstance3 = instance3.copy(reservation = Some(Reservation(Nil, Reservation.State.Launched)))
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala]: Isn't the reservation here on the instance and not the task as the tests says?

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+      verify(instanceRepository, once).ids()
+      verify(instanceRepository, once).store(targetInstance)
+      verify(instanceRepository, once).store(targetInstance2)
+      verify(instanceRepository, never).store(targetInstance3)
+    }
+  }
+
+  private class Fixture {
+
+    val now = Timestamp.now()
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala]: Please use a fixed time.

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+    implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala]: Why is this required here when the migration has a resolver as well.

File: src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala (hunk 15 above)
Comment (meln1k) [reply src/test/scala/mesosphere/marathon/storage/migration/MigrationTo160Test.scala]: This is the expected result.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 13 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Yes

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 14 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: mapConcat unwraps list and emits its elements downstream one by one

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 12 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: This is an implicit parameter for zk store

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 10 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: yep, but here it doesn't really matters since the depth of the call stack is <= 5

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 11 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: We will skip the migration then

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 12 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Don't we have this defined already?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 11 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: Hm, shouldn't this be an error instead?

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 14 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: What do you think about making these to lambdas helper methods? This makes reading and testing much simpler.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 12 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: This one is for operating with `JsValue`s.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+        Some(reservationJson.as[Reservation] -> instance)
+      } getOrElse {
+        None
+      }
+    }
+
+    def checkExistingReservationAndUpdate(reservation: Reservation, instance: Instance): Option[Instance] = {
+      instance.reservation match {
+        case Some(_) =>
+          //do nothing in case instance already contains some reservations, we don't want to not overwrite existing data
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: nit: a space after `//`

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+          None
+        case None =>
+          //no reservation on the instance level, updating instance with provided reservation
```
Comment (ichernetsky) [inline src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: The same here.

File: src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala (hunk 11 above)
Comment (ichernetsky) [reply src/main/scala/mesosphere/marathon/storage/migration/MigrationTo160.scala]: If ZK is not used as a state storage, no migration is needed.

**Comment by ichernetsky** [comment]
Tested manually. Looks good. Merging.
